### PR TITLE
[NHWC CUDNN CONV]Update cudnn convolution memory_format behavior

### PR DESCRIPTION
**Pull request by jjsjann123**
1. Allows both the memory_format of weight & input to dictate the output
memory_format.
2. Provides utility function to recursively convert memory_format of Conv2d and
ConvTranspose2d layers. This allows easy model conversion and ensures that lost
memory_format through incompatible layers could be restored at Convolution-like
layer, where significant performance boost is expected on later generation CUDA
devices.

### CODE REVIEW

**Comment by jjsjann123** [comment]
CC'ing @ptrblck @ngimel for visibility.

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 5b876da7:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=5b876da7c679512220d4a8957095292323be3ade).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4314299) pytorch_linux_xenial_cuda10_1_cudnn7_py3_NO_AVX_NO_AVX2_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89271227) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89271227)) :snowflake:
<details>
<summary>
<code>Jan 22 08:27:23 AssertionError: tensor(2.00272e-05, device='cuda:0', dtype=torch.float32) not less than or equal to 2e-05 :</code>
</summary>

```
Jan 22 08:27:23   File "test_nn.py", line 9989, in test_embedding_bag_device 
Jan 22 08:27:23     self._test_EmbeddingBag(device, 'mean', True, dtype, test_backward=test_backward) 
Jan 22 08:27:23   File "test_nn.py", line 9955, in _test_EmbeddingBag 
Jan 22 08:27:23     self._test_EmbeddingBag_vs_Embedding(N, D, B, L, **kwargs) 
Jan 22 08:27:23   File "test_nn.py", line 9823, in _test_EmbeddingBag_vs_Embedding 
Jan 22 08:27:23     self.assertEqual(es_weight_grad, e.weight.grad, needed_prec) 
Jan 22 08:27:23   File "/var/lib/jenkins/workspace/test/common_utils.py", line 859, in assertEqual 
Jan 22 08:27:23     assertTensorsEqual(x, y) 
Jan 22 08:27:23   File "/var/lib/jenkins/workspace/test/common_utils.py", line 829, in assertTensorsEqual 
Jan 22 08:27:23     self.assertLessEqual(max_err, prec, message) 
Jan 22 08:27:23 AssertionError: tensor(2.00272e-05, device='cuda:0', dtype=torch.float32) not less than or equal to 2e-05 :  
Jan 22 08:27:23  
Jan 22 08:27:23 ---------------------------------------------------------------------- 
Jan 22 08:27:23 Ran 1591 tests in 812.582s 
Jan 22 08:27:23  
Jan 22 08:27:23 FAILED (failures=1, skipped=66, expected failures=4) 
Jan 22 08:27:23  
Jan 22 08:27:23 Generating XML reports... 
Jan 22 08:27:24 Traceback (most recent call last): 
Jan 22 08:27:24   File "test/run_test.py", line 456, in <module> 
Jan 22 08:27:24     main() 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by jjsjann123** [comment]
Looks to be conflicting some upstream changes in `test_nn.py`.
I'll update after initial review.

File: test/test_nn.py [hunk 1]
```diff
@@ -10532,6 +10540,27 @@ def test_conv_cudnn_mismatch_memory_format(self, device):
             self._test_conv_cudnn_nhwc_nchw(nn.Conv2d, n, c, h, w, k, filter_size, device)
             self._test_conv_cudnn_nhwc_nchw(nn.ConvTranspose2d, n, c, h, w, k, filter_size, device)
 
+    @onlyCUDA
```
Comment (VitalyFedyunin) [inline test/test_nn.py]: I think we need to clearly identify cases when usage of `convert_conv2d_weight_memory_layout` is better than doing `model = model.to(memory_format=torch.channels_last)` as this particular case not covering it.

**Comment by VitalyFedyunin** [comment]
Please rebase, I want to test it with cuda9, as I had issues with it trying to introduce similar change.

**Comment by jjsjann123** [comment]
Resolved conflicts in added tests.

File: test/test_nn.py (hunk 1 above)
Comment (jjsjann123) [reply test/test_nn.py]: Do we just want something that explains what `convert_conv2d_weight_memory_layout` does?

I'm copying the message from slack over:
> Calling `model.to(memory_format=torch.channels_last)` is more aggressive.Layers that has 4d tensor as weight may or may not get speedup converting weights to NHWC. One thing we (relatively) know is that NHWC transpose for convolution is beneficial and hopefully will justify the transpose if it is actually required.
> Outside of this context, I don't know if there is another layer with 4d weight that would be affected by .to (a pointwise layer with a 4d weight would be beneficial to be in the same memory_format as the input is.)

I'll get a cleaner text and throw them under the docString of `convert_conv2d_weight_memory_layout`.

Comment (VitalyFedyunin) [reply test/test_nn.py]: Looking forward for cleaner text. Everything else looks good for me.

**Comment by VitalyFedyunin** [comment]
Please ping me when it is good to go! Thanks.

**Comment by jjsjann123** [comment]
@VitalyFedyunin Thanks for keeping an eye on this.
I was waiting for tests to pass. Should be good to go now.

**Review by VitalyFedyunin — APPROVED** [review]

**Pull request closed without merging**